File: linkedlist/Q24.java
public class Q24 {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(0, head), ptr = dummy;
        while (ptr.next != null && ptr.next.next != null) {
            ListNode first = ptr.next, second = ptr.next.next;
            first.next = second.next;
            second.next = first;
            ptr.next = second;
            ptr = first;
        }
        return dummy.next;
    }

    public ListNode swapPairs2(ListNode oldHead) {
        if (oldHead == null || oldHead.next == null) return oldHead;

        ListNode newHead = oldHead.next;
        oldHead.next = swapPairs2(oldHead.next.next);
        newHead.next = oldHead;
        return newHead;
    }
}

File: linkedlist/Q24.py
# https://leetcode.com/problems/swap-nodes-in-pairs/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        ptr = dummy = ListNode(0, head)
        while ptr.next and ptr.next.next:
            first = ptr.next
            second = first.next

            first.next = second.next
            second.next = first
            ptr.next = second
            ptr = first
        return dummy.next
